pretyping/constr_matching.ml

```diff
@@ -263,7 +263,7 @@ let matches_core env sigma allow_bound_rels (binding_vars, pat) c =
 
   | PVar v1, Var v2 when Id.equal v1 v2 -> subst
 
-  | PRef ref, _ when EConstr.isRefX env sigma ref t -> subst
+  | PRef ref, (Const _ | Ind _ | Construct _) when EConstr.isRefX env sigma ref t -> subst
```

Comment:
ref can't be a VarRef?

Reply:
No, this is tested above actually. I could expand isRefX completely if you wish so, it would prevent this kind of confusing code.

Reply:
(Also, I really hate the `where` clauses, they're a great way to write non-robust pattern-matching that is in addition slow. I could just do that in a subsequent PR which removes the when clauses from this code.)